Fix parsing of procfs stat files when comm name contains spaces

stephentoub opened the pull request:
If a process or thread name contains a space, our procfs stat file parsing logic breaks as it sees the space in the name as a separator between elements.  This in turn causes Process.GetProcesses() and other such functions that rely on these stat files to fail with InvalidDataExceptions.  This commit fixes it by adding knowledge of '(' and ')' groupings to the parser.

Fixes https://github.com/dotnet/corefx/issues/12755
cc: @ellismg, @ianhays, @mellinoe, @AndreyAkinshin

mellinoe commented:
The changes look good to me, overall. I wanted to point something out that I found in some brief testing I did. I see there's some tests for files with parentheses in their names, but when I actually create a program with parentheses in the name, I get weird output from `/proc/<pid>/stat`. It doesn't print out the entire name, and often has mismatched parentheses. For example, I created a program named `executable-(name)`, and it shows up in `/proc/<pid>/stat` as `(executable-(nam)`. Did you encounter this kind of behavior? I think that it is also easy to craft a name with a bunch of "right parentheses", which I believe would trigger an exception in the parser.

mellinoe commented on src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/StringParsingHelpers.Connections.cs:
```diff
@@ -139,7 +139,7 @@ public static IPEndPoint[] ParseActiveUdpListenersFromFiles(string udp4File, str
         internal static TcpConnectionInformation ParseTcpConnectionInformationFromLine(string line)
         {
-            StringParser parser = new StringParser(line, ' ', true);
+            StringParser parser = new StringParser(line, ' ', skipEmpty: true);
```
You could also just switch the optional parameters in the overload (make the new one last). Up to you.

mellinoe commented:
To expand on the last point, I created a file called 

`I)mN)o)tEvenTrying)ToMatch)Parentheses)`

and it shows up in `/proc/<pid>/stat` as

`(I)mN)o)tEvenTry)`

which seems bizarre to me.

stephentoub commented:
> For example, I created a program named executable-(name), and it shows up in /proc/<pid>/stat as (executable-(nam)

Just to be clear, you're not using the parsing behavior, and just commenting on what's showing up in the files in procfs, right?  I've not seen the behavior you're referring to.

> To expand on the last point

Hmm, I'm not sure how to handle this.  Seems like maybe I just remove the throw I added for mismatched parens in order to try to be a bit more forgiving.  We could also eat errors we get from parsing and just don't include processes/threads in the result if there's parsing errors.

mellinoe commented:
From the few tests I did, it looks like, if the "name" field in `proc/<pid>/stat` has any parentheses, then it is always "bounded" by matching parentheses. I don't think any of the other fields in the file can ever have parentheses, so I think we may be able to just assume that, if there are any parentheses, then the name is whatever is between the outer matching pair of parentheses. If that assumption doesn't hold up, I'm not sure what to do.

stephentoub commented:
> assume that, if there are any parentheses, then the name is whatever is between the outer matching pair of parentheses

Ok, I'll do that.

stephentoub commented:
Thanks for the feedback, @mellinoe. I've changed it to add a dedicated method to StringParser for parsing the next entry that's expected to be surrounded in the only set of top-level parentheses in the string.  I also added a few more test cases based on our chat.
